token-lending/program/src/dex_market.rs

```diff
@@ -0,0 +1,304 @@
+                let leaf_node = match dex_market_orders.side {
+                    Side::Bid => dex_market_orders
+                        .heap
+                        .as_mut()
+                        .and_then(|heap| heap.remove_max()),
+                    Side::Ask => dex_market_orders
+                        .heap
+                        .as_mut()
+                        .and_then(|heap| heap.remove_min()),
+                }?;
+
+                Some(Order {
+                    price: leaf_node.price().get(),
+                    quantity: leaf_node.quantity(),
+                })
+            }
+            Orders::Cached(orders) => orders.pop_front(),
+            _ => None,
+        }
+    }
+}
+
+/// Trade simulator
+pub struct TradeSimulator<'a, 'b: 'a> {
+    input_currency: Currency,
+    input_lots: u64,
+    output_lots: u64,
+    orders: Orders<'a, 'b>,
+}
+
+impl<'a, 'b: 'a> TradeSimulator<'a, 'b> {
+    /// Create a new TradeSimulator
+    pub fn new(
+        dex_market_info: &AccountInfo,
+        dex_market_orders: &AccountInfo,
+        memory: &'a AccountInfo<'b>,
+        input_reserve: &Reserve,
+        role: Role,
+    ) -> Result<Self, ProgramError> {
+        let dex_market = DexMarket::new(dex_market_info);
+        let dex_market_orders = DexMarketOrders::new(&dex_market, dex_market_orders, memory)?;
+        let (input_currency, input_lots, output_lots) =
+            if input_reserve.liquidity_mint == dex_market.base_mint {
+                (Currency::Base, dex_market.base_lots, dex_market.quote_lots)
+            } else {
+                (Currency::Quote, dex_market.quote_lots, dex_market.base_lots)
+            };
+
+        let order_book_side = match (role, input_currency) {
+            (Role::Taker, Currency::Base) => Side::Ask,
+            (Role::Taker, Currency::Quote) => Side::Bid,
+            (Role::Maker, Currency::Base) => Side::Bid,
+            (Role::Maker, Currency::Quote) => Side::Ask,
+        };
+
+        if order_book_side != dex_market_orders.side {
+            return Err(LendingError::DexInvalidOrderBookSide.into());
+        }
+
+        Ok(Self {
+            input_currency,
+            input_lots,
+            output_lots,
+            orders: Orders::DexMarket(dex_market_orders),
+        })
+    }
+
+    /// Simulate a trade
+    pub fn simulate_trade(
+        &mut self,
+        amount: Decimal,
+        cache_orders: bool,
+    ) -> Result<Decimal, ProgramError> {
+        let input_quantity: Decimal = amount / self.input_lots;
+        let output_quantity = self.exchange_with_order_book(cache_orders, input_quantity)?;
+        Ok(output_quantity * self.output_lots)
+    }
+
+    /// Calculate output quantity from input using order book depth
+    fn exchange_with_order_book(
+        &mut self,
+        cache_orders: bool,
+        mut input_quantity: Decimal,
+    ) -> Result<Decimal, ProgramError> {
+        let mut output_quantity = Decimal::zero();
```

Comment:
This relies on a few core assumptions -- that (for most cases) we will only call this at most twice, and that the second call will be with a value less than or equal to the first call.
* On the first call / filling the cache, we stop once we get enough orders to fill that amount, so on the second call with `cache_orders = false`, we can panic if we go too far.  That's fine in most cases I would imagine.
* If we call more than once with `cache_orders = true`, the cache will end up in an inconsistent state unless we use the same exact orders as on the first call.

Could you document this behavior and add a check to see that we aren't calling with a cache already?

Reply:
Done